### ISSUE TITLE
Issue with conversion of class 'labelled' variables to class 'factor'

### ISSUE THREAD

**Issue by thijsbenschop** [sdcTools/sdcMicro#171]
Hi,

for further testing of the GUI with different data, Cathrine has prepared a Honduras survey dataset in several data formats. The datasets are available in the the shared folder on Box: /Data/Honduras test data. While testing with these data, I have experienced problems with the conversion of variables of class labelled imported with the help of read_xxx function of the haven package to class factor. With the option "values" for the argument 'levels', all numeric values are lost. Trying other options yield better results for this particular variable (H65). Are there other instances where "default" performs worse than "default"? "default" seems to use levels, unless they aren't available, then it uses the values. If there are no cases where this causes troubles, I would suggest to change to "default". The corresponding code is available in lines 420-431 of GUIfunctions.R. I add the code to illustrate the problem based on the different data formats. The problem occurs with the SPSS and STATA files.
```
> # SAS 
> a <- read_sas("/Honduras/HND_2012_L2L.sas7bdat")
> class(a$H65)
[1] "numeric"
> table(a$H65)

   200    400    500    800    900   1000   1200   1500   1600   1800   1900   2000   2500   2700   2850   3000   4000   4800   5000 
    11      6     11     22     15     23      3     10      3      8     27     55     14      3      2     25     17      5      5 
  5800   6000   9000  10000  13000  50000 999998 
     2     24      4      7      6      4     38 
> 
> # SPSS
> b <- read_spss("/Honduras/HND_2012_L2L_spss_windows.sav")
> class(b$H65)
[1] "labelled"
> table(b$H65)

   200    400    500    800    900   1000   1200   1500   1600   1800   1900   2000   2500   2700   2850   3000   4000   4800   5000 
    11      6     11     22     15     23      3     10      3      8     27     55     14      3      2     25     17      5      5 
  5800   6000   9000  10000  13000  50000 999998 
     2     24      4      7      6      4     38 
> table(b$H65)

   200    400    500    800    900   1000   1200   1500   1600   1800   1900   2000   2500   2700   2850   3000   4000   4800   5000 
    11      6     11     22     15     23      3     10      3      8     27     55     14      3      2     25     17      5      5 
  5800   6000   9000  10000  13000  50000 999998 
     2     24      4      7      6      4     38 
> # After reading in the data, class labelled is converted to factor (lines 419-430 in GUIfunctions.R)
> table(as_factor(b$H65, levels="values")) # as in GUI, numeric values are lost

     0 999997 999998 999999 
     0      0     38      0 
> table(as_factor(b$H65, levels="default")) # other option

                 0 L                  200                  400                  500                  800                  900 
                   0                   11                    6                   11                   22                   15 
                1000                 1200                 1500                 1600                 1800                 1900 
                  23                    3                   10                    3                    8                   27 
                2000                 2500                 2700                 2850                 3000                 4000 
                  55                   14                    3                    2                   25                   17 
                4800                 5000                 5800                 6000                 9000                10000 
                   5                    5                    2                   24                    4                    7 
               13000                50000    999,997 L or more Don't Know/Not Known  Refused/No response 
                   6                    4                    0                   38                    0 
> table(as_factor(b$H65, levels="label")) # other option, doesn't work

                 0 L    999,997 L or more Don't Know/Not Known  Refused/No response 
                   0                    0                   38                    0 
> table(as_factor(b$H65, levels="both")) # other option

                      [0] 0 L                           200                           400                           500 
                            0                            11                             6                            11 
                          800                           900                          1000                          1200 
                           22                            15                            23                             3 
                         1500                          1600                          1800                          1900 
                           10                             3                             8                            27 
                         2000                          2500                          2700                          2850 
                           55                            14                             3                             2 
                         3000                          4000                          4800                          5000 
                           25                            17                             5                             5 
                         5800                          6000                          9000                         10000 
                            2                            24                             4                             7 
                        13000                         50000    [999997] 999,997 L or more [999998] Don't Know/Not Known 
                            6                             4                             0                            38 
 [999999] Refused/No response 
                            0 
> 
> #STATA
> c <- read_dta("/Honduras/HND_2012_L2L.dta")
> class(c$H65)
[1] "labelled"
> table(c$H65)

   200    400    500    800    900   1000   1200   1500   1600   1800   1900   2000   2500   2700   2850   3000   4000   4800   5000 
    11      6     11     22     15     23      3     10      3      8     27     55     14      3      2     25     17      5      5 
  5800   6000   9000  10000  13000  50000 999998 
     2     24      4      7      6      4     38 
> # After reading in the data, class labelled is converted to factor (lines 419-430 in GUIfunctions.R)
> table(as_factor(c$H65, levels="values")) # as in GUI, numeric values are lost

     0 999997 999998 999999 
     0      0     38      0 
> table(as_factor(c$H65, levels="default")) # other option

                 0 L                  200                  400                  500                  800                  900 
                   0                   11                    6                   11                   22                   15 
                1000                 1200                 1500                 1600                 1800                 1900 
                  23                    3                   10                    3                    8                   27 
                2000                 2500                 2700                 2850                 3000                 4000 
                  55                   14                    3                    2                   25                   17 
                4800                 5000                 5800                 6000                 9000                10000 
                   5                    5                    2                   24                    4                    7 
               13000                50000    999,997 L or more Don't Know/Not Known  Refused/No response 
                   6                    4                    0                   38                    0 
> table(as_factor(c$H65, levels="label")) # other option, doesn't work

                 0 L    999,997 L or more Don't Know/Not Known  Refused/No response 
                   0                    0                   38                    0 
> table(as_factor(c$H65, levels="both")) # other option

                      [0] 0 L                           200                           400                           500 
                            0                            11                             6                            11 
                          800                           900                          1000                          1200 
                           22                            15                            23                             3 
                         1500                          1600                          1800                          1900 
                           10                             3                             8                            27 
                         2000                          2500                          2700                          2850 
                           55                            14                             3                             2 
                         3000                          4000                          4800                          5000 
                           25                            17                             5                             5 
                         5800                          6000                          9000                         10000 
                            2                            24                             4                             7 
                        13000                         50000    [999997] 999,997 L or more [999998] Don't Know/Not Known 
                            6                             4                             0                            38 
 [999999] Refused/No response 
                            0 
> 
> # CSV
> d <- read.table("/Honduras/HND_2012_L2L.csv", header = TRUE, sep = ",")
> class(d$H65)
[1] "integer"
> table(d$H65)

   200    400    500    800    900   1000   1200   1500   1600   1800   1900   2000   2500   2700   2850   3000   4000   4800   5000 
    11      6     11     22     15     23      3     10      3      8     27     55     14      3      2     25     17      5      5 
  5800   6000   9000  10000  13000  50000 999998 
     2     24      4      7      6      4     38 
> 
```

**Comment by bernhard-da** [comment]
try the fix in branch `issue_171`. if the change to `levels="default"` helps with these kind of input data, I'll merge it.

**Comment by thijsbenschop** [comment]
Thanks for making this change. I just tried and as expected it works better for this particular variable (H65). As you mentioned in your email, it takes the labels also if they are long, which is a disadvantage. I wonder though why with 'values' the numeric values are all coded to NA. I would expect it to take these values and the values of the four remaining levels too.

I'm not sure what's the best option here. Matthew, Cathrine, what do you think?

**Comment by bernhard-da** [comment]
there is also some discussion on this issue [here](https://github.com/tidyverse/haven/issues/81).

**Comment by thijsbenschop** [comment]
Thanks for pointing at this tread!

**Comment by cmachingauta** [comment]
Thanks Thijs for raising this. I experienced the same issue here, and Berhnard thanks for sharing the informative thread. It does look like changing to levels="default" would indeed take care of this particular problem.

**Comment by bernhard-da** [comment]
merged with commit bd2f6c4dfe57809c055db5032d07661eb3e70f73
